src/pygeoprocessing/routing/routing.pyx

```diff
@@ -2960,3 +3006,1298 @@ def _is_raster_path_band_formatted(raster_path_band):
+                    boundary_list)
+                n_left_turns = 0
+            elif left_in:
+                # step forward
+                x_l += D8_XOFFSET[edge_dir]
+                y_l += D8_YOFFSET[edge_dir]
+                # the pixel moves forward
+                boundary_list.append((x_l, y_l))
+                n_left_turns = 0
+            else:
+                # turn left
+                edge_side = edge_dir
+                edge_dir = (edge_side + out_dir_increase) % 8
+                n_left_turns += 1
+
+            if _is_close(x_p, x_first) and _is_close(y_p, y_first):
+                # met the start point so we completed the watershed loop
+                break
+
+        watershed_feature = ogr.Feature(watershed_layer.GetLayerDefn())
+        watershed_polygon = ogr.Geometry(ogr.wkbPolygon)
+        watershed_polygon.AddGeometry(watershed_boundary)
+        watershed_feature.SetGeometry(watershed_polygon)
+        watershed_feature.SetField('stream_fid', stream_fid)
+        watershed_layer.CreateFeature(watershed_feature)
+
+        # this loop fills in the raster at the boundary, done at end so it
+        # doesn't interfere with the loop return to think the cells are no
+        # longer in the watershed
+        for boundary_x, boundary_y in boundary_list:
+            discovery_managed_raster.set(boundary_x, boundary_y, -1)
+    watershed_layer.CommitTransaction()
+    watershed_layer = None
+    watershed_vector = None
+    discovery_managed_raster.close()
+    finish_managed_raster.close()
+    #shutil.rmtree(workspace_dir)
+    LOGGER.info(
+        '(calculate_subwatershed_boundary): watershed building 100% complete')
+
+
+cdef void _diagonal_fill_step(
+        int x_l, int y_l, int edge_dir,
+        long discovery, long finish,
+        _ManagedRaster discovery_managed_raster,
+        long discovery_nodata, boundary_list):
+    """Fill diagonal that are in the watershed behind the new edge.
+
+    Used as a helper function to mark pixels as part of the watershed
+    boundary in one step if they are diagonal and also contained within the
+    watershed. Prevents a case like this:
+
+    iii
+    ii1
+    i1o
+
+    Instead would fill the diagonal like this:
+
+    iii
+    i11
+    i1o
+
+    Args:
+        x_l/y_l (int): leading coordinate of the watershed boundary
+            edge.
+        edge_dir (int): D8 direction that points which direction the edge
+            came from
+        discovery/finish (long): the discovery and finish time that defines
+            whether a pixel discovery time is inside a watershed or not.
+        discovery_managed_raster (_ManagedRaster): discovery time raster
+            x/y gives the discovery time for that pixel.
+        discovery_nodata (long): nodata value for discovery raster
+        boundary_list (list): this list is appended to for new pixels that
+            should be neighbors in the fill.
+
+    Return:
+        None.

```

Comment:
There's no return value, but there is a side effect.  This is a CDEF'd function so probably no one will actually see this unless they're reading the source, but should the side effect be recorded here?

Reply:
I think I do document that side effect in the 'args' as "boundary_list (list): this list is appended to for new pixels that should be neighbors in the fill." Is that okay?